File: backend/transfer.py
import json

import pymongo
from pymongo.errors import DuplicateKeyError

old_db = pymongo.MongoClient("mongodb://localhost:8181/").dclub

new_db = pymongo.MongoClient("mongodb://localhost:27017/").embedg

new_docs = []

for doc in old_db.messages.find({}):
    new_doc = {
        "_id": str(doc["_id"]),
        "owner_id": str(doc["user_id"]),
        "created_at": doc["last_updated"],
        "updated_at": doc["last_updated"],
        "name": doc["name"],
        "description": None,
        "payload_json": json.dumps({
            "username": doc["json"].get("username") or None,
            "avatar_url": doc["json"].get("avatar_url") or None,
            "content": doc["json"].get("content") or None,
            "embeds": [
                {
                    "description": embed.get("description") or None,
                    "timestamp": embed.get("timestamp") or None,
                    "title": embed.get("title") or None,
                    "url": embed.get("url") or None,
                    "color": embed.get("color") or None,
                    "fields": embed.get("fields", []),
                    "author": {
                        "name": embed["author"].get("name") or None,
                        "url": embed["author"].get("url") or None,
                        "icon_url": embed["author"].get("icon_url") or None,
                    } if (embed.get("author") and (
                                embed["author"].get("name") or embed["author"].get("url") or embed["author"].get(
                            "icon_url"))) else None,
                    "footer": {
                        "text": embed["footer"].get("text") or None,
                        "icon_url": embed["footer"].get("icon_url") or None,
                    } if (embed.get("footer") and (
                            embed["footer"].get("text") or embed["footer"].get("icon_url"))) else None,
                    "image": {
                        "url": embed["image"].get("url") or None,
                    } if (embed.get("image") and embed["image"].get("url")) else None,
                    "thumbnail": {
                        "url": embed["thumbnail"].get("url") or None,
                    } if (embed.get("thumbnail") and embed["thumbnail"].get("url")) else None,
                }
                for embed in doc["json"].get("embeds", [])
            ]
        })
    }
    new_docs.append(new_doc)

try:
    new_db.messages.insert_many(new_docs, ordered=False)
except DuplicateKeyError:
    pass

File: embedg-server/db/postgres/transfer/transfer.go
package transfer

import (
	"context"
	"database/sql"
	"encoding/json"
	"fmt"
	"log"
	"time"

	"github.com/lib/pq"
	"github.com/merlinfuchs/discordgo"
	"github.com/merlinfuchs/embed-generator/embedg-server/actions"
	"github.com/merlinfuchs/embed-generator/embedg-server/db/postgres"
	"github.com/redis/go-redis/v9"
	"github.com/spf13/viper"
	"go.mongodb.org/mongo-driver/bson"
	"go.mongodb.org/mongo-driver/mongo"
	"go.mongodb.org/mongo-driver/mongo/options"
)

type OldSavedMessage struct {
	ID          string    `bson:"_id"`
	OwnerID     string    `bson:"owner_id"`
	CreatedAt   time.Time `bson:"created_at"`
	UpdatedAt   time.Time `bson:"updated_at"`
	Name        string    `bson:"name"`
	Description string    `bson:"description"`
	PayloadJSON string    `bson:"payload_json"`
}

type OldUsers struct {
	ID            string `bson:"_id"`
	Username      string `bson:"username"`
	Discriminator string `bson:"discriminator"`
	Avatar        string `bson:"avatar"`
}

func TransferDB() {
	pg := postgres.NewPostgresStore()

	redisDB := redis.NewClient(&redis.Options{
		Addr: "localhost:6379",
	})

	mongoURL := viper.GetString("mongo.url")
	if mongoURL == "" {
		mongoURL = "mongodb://localhost:27017"
	}

	mongo, err := mongo.Connect(context.TODO(), options.Client().ApplyURI(mongoURL))
	if err != nil {
		panic(err)
	}

	coll := mongo.Database("embedg").Collection("messages")

	var skip int64 = 0

	cursor, err := coll.Find(context.TODO(), bson.M{
		// "owner_id": "386861188891279362",
	}, options.Find().SetBatchSize(100).SetSkip(skip).SetSort(bson.M{"updated_at": -1}))
	if err != nil {
		panic(err)
	}

	session, err := discordgo.New("Bot " + viper.GetString("discord.token"))
	if err != nil {
		panic(err)
	}

	i := skip
	for cursor.Next(context.Background()) {
		var result OldSavedMessage
		if err := cursor.Decode(&result); err != nil {
			log.Fatal(err)
		}

		_, err := pg.Q.GetUser(context.TODO(), result.OwnerID)
		if err != nil {
			if err != sql.ErrNoRows {
				panic(err)
			}

			raw, err := redisDB.Get(context.Background(), "users:"+result.OwnerID).Result()
			if err != nil {
				if err != redis.Nil {
					panic(err)
				}

				user, err := session.User(result.OwnerID)
				if err != nil {
					panic(err)
				}

				fmt.Println("fetched user")

				pg.Q.UpsertUser(context.TODO(), postgres.UpsertUserParams{
					ID:            user.ID,
					Name:          user.Username,
					Discriminator: user.Discriminator,
					Avatar:        sql.NullString{String: user.Avatar, Valid: user.Avatar != ""},
				})
			} else {
				var user OldUsers
				err = json.Unmarshal([]byte(raw), &user)
				if err != nil {
					panic(err)
				}

				fmt.Println("got user from redis")

				pg.Q.UpsertUser(context.TODO(), postgres.UpsertUserParams{
					ID:            user.ID,
					Name:          user.Username,
					Discriminator: user.Discriminator,
					Avatar:        sql.NullString{String: user.Avatar, Valid: user.Avatar != ""},
				})
			}
		}

		var msgData actions.MessageWithActions
		err = json.Unmarshal([]byte(result.PayloadJSON), &msgData)
		if err != nil {
			fmt.Println("Failed to unmarshal message", result.ID)
			continue
		}

		if msgData.Content == "" && len(msgData.Components) == 0 && len(msgData.Embeds) == 0 {
			continue
		}

		newEmbeds := make([]*discordgo.MessageEmbed, 0)
		for _, embed := range msgData.Embeds {
			if embed.Footer != nil || embed.Author != nil || embed.Title != "" || embed.Description != "" || len(embed.Fields) != 0 || embed.Image != nil || embed.Thumbnail != nil {
				newEmbeds = append(newEmbeds, embed)
			}
		}

		msgData.Embeds = newEmbeds

		rawMsg, err := json.Marshal(&msgData)
		if err != nil {
			panic(err)
		}

		_, err = pg.Q.InsertSavedMessage(context.TODO(), postgres.InsertSavedMessageParams{
			ID:          result.ID,
			CreatorID:   result.OwnerID,
			GuildID:     sql.NullString{},
			UpdatedAt:   result.UpdatedAt,
			Name:        result.Name,
			Description: sql.NullString{Valid: result.Description != "", String: result.Description},
			Data:        json.RawMessage(rawMsg),
		})
		if err != nil {
			if (err.(*pq.Error)).Code == "23505" {
				continue
			}
			panic(err)
		}

		i++
		fmt.Println(i)
	}
}
